src/core.jl

```diff
@@ -129,7 +183,9 @@ function jnew(T::Symbol, argtypes::Tuple, args...)
     if jmethodId == C_NULL
         throw(JavaCallError("No constructor for $T with signature $sig"))
     end
-    return  _jcall(metaclass(T), jmethodId, jnifunc.NewObjectA, JavaObject{T}, argtypes, args...)
+    result = _jcall(metaclass(T), jmethodId, jnifunc.NewObjectA, JavaObject{T}, argtypes, args...)
+    deletelocals()
```

Comment:
@zot Why is this necessary? Is the garbage collector not working for the local references for some reason?

Reply:
Localrefs are raw pointers / handles into the JVM so the Julia gc does not interact with them and we're not necessarily holding onto them with a finalizable object, just a Ptr{nothing} (Julia objects use global refs not local refs).

I think that local references created outside a call will not be collected until you explicitly deallocate them. It's possible, however, that the JVM blows away all localrefs at the end of each call so I'm not positive about this.

Reply:
Thank you, @zot . Sorry for being a bit curt. It was getting late.

I've quoted the documentation below.

https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#referencing_java_objects

> In most cases, the programmer should rely on the VM to free all local references after the native method returns. However, there are times when the programmer should explicitly free a local reference. Consider, for example, the following situations:
> * A native method accesses a large Java object, thereby creating a local reference to the Java object. The native method then performs additional computation before returning to the caller. The local reference to the large Java object will prevent the object from being garbage collected, even if the object is no longer used in the remainder of the computation.
> * A native method creates a large number of local references, although not all of them are used at the same time. Since the VM needs a certain amount of space to keep track of a local reference, creating too many local references may cause the system to run out of memory. For example, a native method loops through a large array of objects, retrieves the elements as local references, and operates on one element at each iteration. After each iteration, the programmer no longer needs the local reference to the array element.
> 

I'm not clear if we are necessarily in one of those special cases. I suppose we should figure how to check if the JVM is doing that job in this context. I'm not clear what would constitute a native method here.

If this is needed, I'm wondering if we should just build it into `_jcall`. Also, is this not needed for methods that return primitives?

Reply:
For example, do we also need to track local references in the other `_jcall`:
https://github.com/JuliaInterop/JavaCall.jl/blob/e7e8700eca9a48b7e47f55ee57a25a479de1ad5b/src/core.jl#L289

Reply:
I think I did build it into _jcall, or maybe I'm misunderstanding you?

Right, primitives don't need to be registered but registerlocal() checks for references before adding it to the list.

Looks like I added the registerlocal() function, all it does is add references to an array so they can be unallocated later.

I'm not sure whether or not _jfield needs this.

Reply:
Regarding `_jcall` we currently have a pattern where you store the result of `_jcall`, call `deletelocals()`, and then return the result such as in this in version of `jcall`.

https://github.com/JuliaInterop/JavaCall.jl/blob/e7e8700eca9a48b7e47f55ee57a25a479de1ad5b/src/core.jl#L205-L214

If we need this, I'm wondering if we should just integrated this pattern into `_jcall` itself. It looks like you have considered this because it is commented out:
https://github.com/JuliaInterop/JavaCall.jl/blob/e7e8700eca9a48b7e47f55ee57a25a479de1ad5b/src/core.jl#L312-L318

Per the documentation above, it does look like the Java garbage collector might clean this up automatically. I would like to do an experiment to verify this.

Even if it does clean this up automatically via the Java GC, there may be some utility to making the manual local referencing cleanup available to the user. If you have any further insight or perhaps have done an experiment on this yourself, please let me know.

Thank you for taking the time to explain. I appreciate it.

Reply:
No problem. My experiments with this were in 2018 and my memory of them is a bit hazy. If I remember something other than the things I recounted already I'll mention it.

Reply:
We should consider using `PushLocalFrame` and `PopLocalFrame` to deal with local references rather than implementing our own management system. I am not sure if we need a global reference for everything.
https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#PushLocalFrame

Reply:
Sure, whatever works. Global references are really only needed for the proxies

We do need to ensure that references returned from the JNI don't get randomly collected before the Julia code has a chance to decide what to do with it so think deleting locals inside of `jcall` would be dangerous.

Reply:
I agree. The Java garbage collector does not seem to functioning for JavaCall code because we are not using `PushLocalFrame` and `PopLocalFrame`. Without those, the Java GC does not know when we are out of the local frame.

Reply:
If release the returned reference inside of `jcall`, Java can GC the return value before Julia has a chance to use it, so we need to manage returned references outside of jcall. Otherwise they won't be safe to use for the code that uses jcall.

It might make sense to provide a function that takes a do block and runs it in between a push and a pop so you can safely run a bunch of jcalls in the block.